README.md
# CSASM -> ASMHUB #
**This is the Code Source repository for the Geocaching Anti-Social Mob Hub application (ASMHUB).**

The app requires multiple mobile devices, each connecting to the same source to determine if they are within a defined proximity radius of any waypoint location stored in a database table for a given Game-Code. The app displays _for the current player_ where they are on a map in relation to the target waypoints, highlighting if they are within one. It also keeps track of how far other players are from each waypoint, and will highlight any occupied waypoint on the map. Once all waypoints are occupied, the app returns hidden REWARD text value (eg: clues/directions to the final container) to each player that is occupying a waypoint.

## TODO ##
- [ ] Shrewsbury Park'n'Ride ASM cache for Mega?

 
## Versioning vN.G.PR ##
Where:
- N = Game Version (1 until a major change)
- G = Number of Production Games (not counting test games)
- PR = Production Code Release version from Heroku
and the Development Code release number is noted in changelog.md document as well.

## Key files ##
[SERVER.JS](/server.js) : the primary engine (server side javascript creates and runs index.ejs) - no visibility of goings on to enduser (can only be seen in heroku console). Must socket.emit to communicate to and from client side javascripts (c.f. socket.on(LOGX))

[INDEX.EJS](/views/pages/index.ejs) : the calling script (client side html) to display stuff  - visible in console (F12).

[CLIENT.JS](/public/js/client.js) : the primary controller (client side javascript called from index.ejs) - visible in console (F12). 

[SHOWMAP.JS](/public/js/showmap.js) : the map code (client side javascript called from index.ejs) - visible in console (F12). 

[REWARD.EJS](/views/pages/reward.ejs) : the final reward display script (client side html) - visible in console (F12).


## LINKS ##
- [Github Pprime1/csasm	](https://github.com/Pprime1/csasm)
- [Heroku Pipeline](https://dashboard.heroku.com/pipelines/8ac5b1b3-7f26-4c08-a927-f54f4c14888f)
- [csasm · Logs | Heroku	](https://dashboard.heroku.com/apps/csasm/logs)
- [Heroku Database page	](https://data.heroku.com/dataclips)
- [Heroku | ASMHUB appsite	](https://asmhub.herokuapp.com/)
- [The Anti-Social Mob Cache](https://www.geocaching.com/geocache/GC9JEH6_the-anti-social-mob?guid=c7d27f31-980c-42ea-bd66-545bb5d77874)
- [asmhub GUIDE](https://enblesp1msgzhmk140wukq-on.drv.tw/asmhub/asmhubGuide.html)


CHANGELOG.md
# HISTORY OF CODE DEVELOPMENT RELEASES #

**PRODUCTION RELEASE v1.1.23**
 - committed 25 July 2022
 - Build Release v1115 created by user pprime@live.com.au 2022-25-06T02:52
 - Production Code Release v23
 - [ ] Limit mapping update logs to once each time postition changes
 - [ ] Added Geolocation error short name primer into console when used
 

**PRODUCTION RELEASE v1.1.22**
 - committed 06 July 2022
 - Build Release v1104 created by user pprime@live.com.au 2022-07-06T04:40
 - Production Code Release v22
- [x] Dependabot security patching - EJS update from 3.1.6 to 3.1.7
- [x] Dependabot security patching - GOT update from 11.8.2 to 11.8.5
- [x] Dependabot security patching - minimist update from 1.2.5 to 1.2.6
- [x] OAuth connector to Heroku reconnected after Github OAuth security issue


**PRODUCTION RELEASE v1.1.21**
 - committed 21 February 2022
 - Build Release v1101 created by user pprime@live.com.au 2022-02-21T04:40
 - Production Code Release v21
- [x] reintroduce validatechecksums = true for the database tables. how to set/reset md5 values? (see https://github.com/Pprime1/csasm/issues/15)
- [x] Email alerts on Game start and Game Reward - only alerting once per minute and only if LogEntries screen is not active
- [x] Create a new game form
- [x] Update versioning to be :  v[Functionality Release].[active Geocache games].[Release number]   {Build number}      is v1.1.21 {1100} currently
- [x] Put a 'Help' button top right of screen to UserGuide
- [x] Improve server log (GCALPHA : 0 of 2 waypoints occupied.) to indicate number of current players
- [x] Upgrade all npm audit componentry
  - [x] update engine.io and socket.io using npm audit fix (see issue https://github.com/Pprime1/csasm/issues/34_)
    - [x] Install Github Desktop and sync a clone of csasm
- [x] reduce further/cleanup the F12 - client console.logs
- [x] Update a LOGTX to show distance from each waypoint on updatemap
- [x] Delay first display-updates until player's location has been determined
  - round(ST_DISTANCE(wp.location, pl.location) * 100000) as "distance" is returning null sometimes (early on)
  - [x] Delay startupmap until player location is known

 
 **PRODUCTION RELEASE v1.0.20**
 - committed 6 February 2022
 - Build Release v1052 created by user pprime@live.com.au 2022-02-05T22:22:22
 - Production Code Release v20
- [x] npm audit ... fix
   - + socket.io@4.4.1
   - added 1 package from 1 contributor, removed 1 package and updated 12 packages in 2.761s
   - fixed 2 of 2 vulnerabilities in 116 scanned packages
- [x] Improve client logging transferred to server console 
  - [x] null distance trap
  - [x] capture form use for when URL param not used
  - [x] remove the Update Circle log spam
  - [x] circle distance once occupying
- [📧] Get emailed alerts on Chosen Game and Game REWARD events ... unreliable?

**PRODUCTION RELEASE v1.0.19**
 - committed 4 February 2022
 - Build Release v1045 created by user pprime@live.com.au 2022-02-03T22:38:50
 - Production Code Release v19
- [👎] reintroduce validatechecksums = true for the database tables. how to set/reset md5 values? (see https://github.com/Pprime1/csasm/issues/15)
- [x] make sure github code is hidden from everyone except Andrew and me. Database is secure, source code is harmless
- [x] shorten console.log for leaving game game-ID code (remove 'game-')
  - 2022-01-28T03:40:15.553001+00:00 app[web.1]: nNb7pC3X4vhRDbl1AAAD left game-GC9JEH6 0 online
  - 2022-01-28T03:40:15.553127+00:00 app[web.1]: All players have left game:  game-GC9JEH6
- [x] Push certain client console.logs to the server for centralised display
- [x] Push certain showmap console.logs to the server for centralised display
- [x] Reduce server console log entries in the ten second cycle
  - 2022-02-02T04:27:03.286916+00:00 app[web.1]: new player: aYDBQZvsPmGR0tqQAAAB
  - 2022-02-02T04:27:03.287646+00:00 app[web.1]: aYDBQZvsPmGR0tqQAAAB connected
  - 2022-02-02T04:27:18.532435+00:00 app[web.1]: Chosen Game is GCBETA : The Anti-Social-Mob cache: Beta test
  - 2022-02-02T04:27:18.532495+00:00 app[web.1]: new game: GCBETA
  - 2022-02-02T04:27:18.533278+00:00 app[web.1]: aYDBQZvsPmGR0tqQAAAB joined game-GCBETA 1 online
  - 2022-02-02T04:27:25.968680+00:00 app[web.1]: Playing game GCBETA
  - 2022-02-02T04:27:25.972407+00:00 app[web.1]: GCBETA requires 2 waypoints to be occupied.
  - 2022-02-02T04:27:25.972453+00:00 app[web.1]: 0 waypoints are currently occupied.
- [x] Can we siphon off the server console.logs for post action review? 
  - [x] Use a logging addon = LogEntries
    - [x] Need to add a valid credit card to Heroku
  - [x] Setup dashboards, tags, filters ... learn to use LogEntries addon
  - [👎] Email alerts on Game start and Game Reward - not happening reliably
- [x] Can we create a spoofing detector? "Anti-Gary mode enabled"
  - [x] accuracy checker for the chrome dev console (exactly 150m)
  - [x] android position.ismockprovider is true ... not sure this is overly reliable, but it's there
  - [👎] lack of movement detector? no FakeGPS simulates movement as well anyway.
  - [x] Gary test - passed, perfect
  

**PRODUCTION RELEASE v1.0.18**
 - committed 19 January 2022
 - Build Release v1000 created by user pprime@live.com.au 2022-01-18T23:21:24.320714+00:00
 - Production Code Release v18
- [x] GC Code input validation to also strip spaces, or at least trailling spaces = use str.trim()
- [x] Why is the Game Code not displaying on the main output? use current-GameID and GameId separately
- [x] Add extra GeoLocation error message/helpnote for users of Facebook Messenger (Android) - use a real browser!
- [x] rerun Testing Day 
  - booked for Friday 14/1/22 16:30
  - [x] Test old iphone
  - [x] Test iphone Go To Top exit button
  - [x] Test landscape on iphone
  - [ ] Test ipad
  - [x] Test Android tablet
  - [x] Test tablets in landscape
  - [x] Test Browser drop outs/reconnect
  - [x] test a refresh does an actual restart
  - [x] create GCBeta (2 waypoints) at QUT 

index.js
const express = require('express')
const path = require('path')
var http = require("http")
const pool = require('./app/database').db_pool;
const PORT = process.env.PORT || 5000

// Express HTTP Server Start
const app = express()

app.use(express.static(path.join(__dirname, 'public')))
  .set('views', path.join(__dirname, 'views'))
  .set('view engine', 'ejs')
  .get('/', (req, res) => res.render('pages/index'))

// Express HTTP Server END

///// WEB SOCKET EXAMPLE

var WebSocket = require("ws")
var WebSocketServer = require("ws").Server

const url = require('url');
const querystring = require('querystring');

var wss = new WebSocketServer({server: server})

const map = new Map();

wss.on('connection', function connection(ws, request) {
	const username = querystring.decode(url.parse(request.url).query).name;
	map.set(username, ws);

	ws.on('message', function incoming(message) {
		let incomingMessage = JSON.parse(message);

		if(incomingMessage.type == "chat") {
			wss.clients.forEach(function each(client) {
			  if (client.readyState === WebSocket.OPEN) {
				client.send(
					JSON.stringify(
						{
							type: 'chat',
							content: `${username}: ${incomingMessage.content}`
						}
					)
				);
			  }
			});
		}
		if(incomingMessage.type == "location") {
			pool.query("UPDATE players set location = ST_GeomFromText('POINT(" + incomingMessage.latitude + " " + incomingMessage.longitude + ")', 3857) WHERE name = $1",
				[username]).catch(error => {
				console.error(error.stack)
			});

			// show places listed in geocache table as long as they're within 50,000 metres of my current location.
			let metres = 50000;

			let qry = `
				SELECT id, description,
				   round(ST_DISTANCE(
						location,
						ST_GeomFromText('POINT(${incomingMessage.latitude} ${incomingMessage.longitude})', 3857)) * 100000) as "distance (m2)"
					from geocaches
					WHERE
						ST_DISTANCE(
						location,
						ST_GeomFromText('POINT(${incomingMessage.latitude} ${incomingMessage.longitude})', 3857)) * 100000 <= ${metres}`;

			pool.query(qry, (error, results) => {
				if (error) {
				  console.error(error.stack);
				} else {
				  ws.send(
					JSON.stringify(
						{
							type: 'distance',
							content: results.rows
						}
					));
				}
			})
		}
	});

	wss.clients.forEach(function each(client) {
	  if (client.readyState === WebSocket.OPEN) {
		client.send(
			JSON.stringify(
				{
					type: 'chat',
					content: `${username} has joined the chat!`
				}
			)
		);
	  }
	});

	ws.on('close', function () {
		pool.query("DELETE FROM players WHERE name = $1", [username]).catch(error => {
			console.error(error.stack)
		});
		map.delete(username);
	});

	pool.query("INSERT INTO players(name) VALUES($1)", [username])
		.catch(error => {
			console.error(error.stack)
			ws.close();
		});

});



/// WEB SOCKET EXAMPLE END


app/database.js
//***DATABASE.js***//
const POSTGRATOR = require('postgrator')
const PG_PARSE = require('pg-connection-string').parse;
const PG_POOL = require('pg').Pool
const path = require('path')

function db(connectionString) {
  return new Promise(resolve => {
    (async function() {
      const postgrator = new POSTGRATOR({
            validateChecksums: true, // Set this to true to force validation which is better security. checks actual md5 hash to that in schematable
            newline: 'CRLF', // Force using 'CRLF' (windows) or 'LF' (unix/mac)
            migrationDirectory: path.join(__dirname, '../postgrator'),
            driver: 'pg',
            ssl: { rejectUnauthorized: false },
            connectionString
      });

      postgrator.on("migration-started", () => console.info("Migration started"));
      postgrator.on("migration-finished", () => console.info("Migration finished"));

      const migrationResult = await postgrator
        .migrate()
        .catch((error) => {
            console.error(`Migration Failure: ${error}`);
            process.exit(2);
        });

        console.info(`${migrationResult.length} migrations completed`);
        var pg_config = PG_PARSE(connectionString);

        // Amend Config to add SSL configuration
        pg_config = {
            ...pg_config,
            ssl: {
                rejectUnauthorized: false
            }
        };

        resolve(new PG_POOL(pg_config));
    }());
  });
}

module.exports = { db };


public/js/client.js
//***CLIENT.js***// 
const socket = io(); // or io("/"), the main namespace
const urlParams = new URLSearchParams(location.search);
var latitude =-27.5;    // defaults to UQ St Lucia
var longitude =153;   // defaults to UQ St Lucia
var is_joined = false; // status for being part of a game
var map_started = false; //set once the map has been started up
var game=null; //current game ID
var game_description = "Clear skies: choose a game to play";
var displaytable =[]; 
var MYID = socket.id; 
var RtnError = null;
var geoOptions = {
  enableHighAccuracy: true,
  timeout: 5000,
  maximumAge: 0
};
var spoof = false; //status for detection of locational spoofing activities
var spoofMsg = "Anti Gary mode enabled. \n This game is a physical location game, please ensure you are physically visiting the locations. If you believe this is a mistaken detection, please contact the game owner.";

for (var entry of urlParams) { 
    game = entry[0]; // only the first URL param is considered as the Game ID code
};   
if (game) { //if started with a URLParam then attempt to join that game ID
    game = game.toUpperCase();
    console.log("Client session called with parameter:", game);
    socket.emit('join-a-game', game, (response) => {         
        $("#game-error").text(response.message); // Set to display an error message underneath form entry field
     }); // emit join-a-game
} else { game = "GC" }; // set default form entry content for if no valid game parameter is provided

function ConvertDEGToDM(deg,dir) {
  var absolute = Math.abs(deg);
  var degrees = Math.floor(absolute);
  var minutesNotTruncated = (absolute - degrees) * 60;
  var minutes = Math.floor(minutesNotTruncated);
  var minutesdecimals = ((absolute - degrees) * 60).toFixed(3);
  var seconds = ((minutesNotTruncated - minutes) * 60).toFixed(2);
  if (dir == 1) {
     	var direction = deg >= 0 ? "N" : "S";
  } else {
      	var direction = deg >= 0 ? "E" : "W";
  }
  return direction + degrees + "° " + minutesdecimals+ "' ";
 }; // Convert DM.MMM

function updatePosition(position) { //changes the player location details:
  latitude = position.coords.latitude; //main variables
  longitude = position.coords.longitude;
  var accuracy = position.coords.accuracy;
  var lat = ConvertDEGToDM(latitude,1); //change display format
  var lon = ConvertDEGToDM(longitude,0);
  var acc = Math.round(accuracy);
  if (acc==150) {spoof=true}; //devtools in client browser uses a fixed accuracy of 150m
  if (position.isFromMockProvider) {spoof=true}; //android option only .... except this doesn't work on FakeGPS?  ?
  $("#current-Lat").text(lat); //index.ejs for display above the table
  $("#current-Lon").text(lon);
  $("#current-Acc").text(acc);
  localStorage.setItem('my_lat', lat); //localstorage for ease of access in reward.ejs
  localStorage.setItem('my_lon', lon);
  localStorage.setItem('my_acc', accuracy);
  socket.emit('location-update', latitude, longitude); //server.js, which updates database for use in display-update later
  console.log("watchposition change", lat,lon);
  if (map_started) {updatemap(latitude,longitude,displaytable)}; //the map display - if the map has been created already
}; // UpdatePosition

function PosError(error) { // handle/display get geolocation errors
    socket.emit('LOGTX',`${socket.id} :-> Geolocation Error ${error.code}`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
    socket.emit('LOGTX',`1=PERMISSION_DENIED; 2=POSITION_UNAVAILABLE; 3=TIMEOUT;`);
    switch (error.code) {
        case error.PERMISSION_DENIED:
            RtnError = "GeoLocation error: User denied the request for Geolocation. \n Please un-block location sharing for this webpage, and then refresh screen to restart.";
            window.alert(RtnError);
            window.open('https://enblesp1msgzhmk140wukq-on.drv.tw/ASMHUB/BrowserSettings.html', '_blank'); // popup in new tab/window
            location.href = "/";
            break;
        case error.POSITION_UNAVAILABLE:
            RtnError = "GeoLocation error: Location information is unavailable \n Please correct and then refresh screen to restart. \n\n If you opened the game from a Facebook messenger link, please re-open it from a full Internet browser app such as Chrome, Safari, Samsung, Firefox or Edge.";
            window.alert(RtnError);
            location.href = "/";
            break;
        case error.TIMEOUT:
            RtnError = "GeoLocation error: The request to get user location timed out. \n Please ensure you are not inside a concrete bunker, and then refresh screen to restart.";
            window.alert(RtnError);
            location.href = "/";
            break;
        default:
            RtnError = "GeoLocation error: An unknown error occurred. \n Please reboot your smartphone and restart.";
            window.alert(RtnError);
            location.href = "/";
            break;
    };
}; // GeoLocation Error handler

socket.io.on("reconnect", () => { // Reconnect if the client has dropped out for some reason but game was in progress
  if (is_joined) { 
      game = $("#gameId").val();
      console.log(`RECONNECTING CLIENT ${ game }`);
      socket.emit('LOGTX',`${socket.id} :-> Reconnecting Client to existing game`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
      socket.emit('join-a-game', game, (response) => {
      $("#game-error").text(response.message); // Set to display an error message underneath form entry field
      }); // emit join-a-game again
  };
});

socket.on("game-join", () => {
   navigator.geolocation.getCurrentPosition(updatePosition, PosError, geoOptions); // First location update attempt, handle errors
   if (RtnError) {console.log("Geolocation Error Status", RtnError);
     socket.emit('LOGTX',`${socket.id} :-> Geolocation Error`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
   } else { 
       $("#lj-startup").hide();
       $("#lj-in-game").show();
       is_joined = true; 
       navigator.geolocation.watchPosition(updatePosition, PosError, geoOptions); //keep updating geolocation when it changes, rather than on a loop
   }; //Get and set Player GeoLocation, and keep watch on it moving
}); // end of GAME-JOIN

socket.on("room-update", (game_id, gamedesc, new_player_count) => {
  is_joined = true;
  $("#lj-startup").hide();
  $("#current-gameId").text(game_id);
  if (gamedesc != 1) { // don't update if this was called by a room-leave command
    localStorage.setItem('game_description', gamedesc); //?
    localStorage.setItem('current_game', game_id); //?
    game_description = gamedesc;
  };
  console.log(new_player_count, "players are joined to", game_id, ":", gamedesc);  
  $("#current-game-player-count").text(new_player_count);
  $("#lj-in-game").show();
}); // end of ROOM-UPDATE

socket.on("display-update", (display_information) => {
  if (spoof) {
      console.log("Spoofing detected");
      socket.emit('LOGTX',`${socket.id} :-> Spoofing detected`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
      window.alert(spoofMsg);
      location.href = "https://www.geocaching.com/help/index.php?pg=kb.chapter&id=141&pgid=46"
  };
  if (display_information[0].distance==null) { //don't do anything if the player's location hasn't been found yet
      socket.emit('LOGTX',`${socket.id} :-> Trying to display with null distances`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
  } else { 
    MYID = socket.id; // this is current player
    displaytable=display_information; //used for updatemap in update position function 
    var DTStamp = new Date(display_information[0].updated_at).toLocaleTimeString('en-GB'); // Last Room update timestamp
    $("#gamedesc").text(game_description);

    // Display game status including any occupied status lines
    var $table = "<table border='1'> <caption>Current Player: " + MYID + " at " + DTStamp + "</caption>"
        $table += "<thead><tr class='table table-primary'><th>Player</th><th>Waypoint</th><th>Radius</th><th>Distance</th></tr></thead><tbody>"
    for (var i = 0; i < display_information.length; i++) {
      //if (display_information[i].distance != null) { // if it is null there is an error somewhere
          if (display_information[i].distance <= display_information[i].radius) {  // For display purposes only, not used for success determination here
              $table += "<tr class='table table-success'>"
          } else {
              $table += "<tr class='table table-light' border='1'>"
          };
          $table += "<td>" + display_information[i].id + "</td>"
          $table += "<td>" + display_information[i].name + "</td>"
          $table += "<td>" + display_information[i].radius + "m</td>"
          $table += "<td>" + display_information[i].distance.toLocaleString() + "m</td></tr>"
     };
    $table += "</tr></tbody></table>";
    $('#displayinfo').empty().append($table);
    localStorage.setItem('display_update', $table); //needed in reward.ejs
    //socket.emit('LOGTX',`${socket.id} :-> Displaying ${JSON.stringify(display_information)}`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
    if (!map_started) {
         startupmap(latitude,longitude,display_information,MYID);
         map_started=true //startup the map, but just the first time we get here;    
    }
  }; //only if distance != null
}); // end of DISPLAY-UPDATE

socket.on("display-reward", (reward_information) => { //if all waypoints are in occupied state, show Success! ONLY SENT TO VALID PLAYERS
  if (!spoof) {
    localStorage.setItem('reward_information', reward_information); // Save Reward Info into Local Storage
    setTimeout( function() {
      location.href = "reward"; // Redirect user to reward page, disconnecting them from game and any session updates.
    }, 100);
  } else {console.log("Reward failed due to spoof attempt");
         socket.emit('LOGTX',`${socket.id} :-> Reward failed due to spoofing attempt`); //clientlogdata should always be in the format of `${socket.id} :-> log message`};
  }
}); // end of DISPLAY-REWARD

// Bind Submit Event for Start Page Game-Joining form.
window.addEventListener("load",function(event) {
  document.querySelector("#gameId").value = game;
  $("#game-error").text(RtnError); //Set to display any error message underneath form entry field
  if (game=="GC") {console.log("No valid game-code supplied, Starting Game Form");}
  if (!is_joined) {$("#lj-startup").show() }; //show the form only if not already joined to a game
      
  $( "#join-game-form" ).on( "submit", function(e) {
     e.preventDefault();
     var game = $("#gameId").val();
     socket.emit('LOGTX',`${socket.id} :-> Form entry used= ${game}`); //clientlogdata should always be in the format of `${socket.id} :-> log message`
     game = game.trim();
     game = game.toUpperCase();
     $("#gameId").text(game);
     socket.emit('join-a-game', game, (response) => {
        $("#game-error").text(response.message); // Set to display any error message underneath form entry field
     }); // emit join-a-game
   }); // end of form
}, false); // end of GAME listener


public/js/mobcache-socketio.js
const socket = io(); // or io("/"), the main namespace

  function ConvertDEGToDM(deg,dir) {
    	var absolute = Math.abs(deg);
    	var degrees = Math.floor(absolute);
    	var minutesNotTruncated = (absolute - degrees) * 60;
    	var minutes = Math.floor(minutesNotTruncated);
    	var minutesdecimals = ((absolute - degrees) * 60).toFixed(3);
    	var seconds = ((minutesNotTruncated - minutes) * 60).toFixed(2);
    	if (dir == 1) {
        	var direction = deg >= 0 ? "N" : "S";
    	} else {
        	var direction = deg >= 0 ? "E" : "W";
    	}
    	return direction + degrees + "° " + minutesdecimals+ "' ";
 }

function updatePosition(position) {
  var latitude = position.coords.latitude;
  var longitude = position.coords.longitude;
  var lat = ConvertDEGToDM(latitude,1);
  var lon = ConvertDEGToDM(longitude,0); //why can't I run this?
    $("#current-Lat").text(lat); // should be lat/lon as formatted. sigh
    $("#current-Lon").text(lon);
  socket.emit('location-update', latitude, longitude);
}

let is_joined = false;

socket.io.on("reconnect", () => {
  if (is_joined) {
      socket.emit('join-a-group', $("#current-group-id").text())
  }
});

socket.on("room-join", () => {
  $("#lj-startup").hide();
  $("#lj-in-game").show();

  navigator.geolocation.getCurrentPosition(updatePosition);
	const interval = setInterval(function() {
		navigator.geolocation.getCurrentPosition(updatePosition);
	}, 10000);
});

socket.on("room-update", (group_id, new_player_count) => {
  is_joined = true;

  $("#lj-startup").hide();
  $("#lj-in-game").show();

  $("#current-group-id").text(group_id);
  $("#current-group-member-count").text(new_player_count);
})

socket.on("room-location-update", (waypoint_information) => {
  console.log(waypoint_information);
  var result= JSON.stringify(waypoint_information,null,2);
  $("#wpinfo").text(result);

  // TODO: Show all waypoints in a table: pass variables from server.js?
	// TODO: Receive from server.js and pass to index.ejs?
        // pl.id WHERE id= '${socket.id}' // this is current player?
        // pl.location
        // pl.updated_at
        
	var wp = waypoint_information;
	
        $("#wpname").text(waypoint_information[0].name);
        $("#wpradius").text(waypoint_information[0].radius);
        $("#distance").text(waypoint_information[0].distance);

	
	// if distance <= wp.radius then set wp.occupied = true // reset to false every room refresh?
  // IF all waypoints have a wp.occupied = true then room-reward is achieved
  
})

socket.on("room-reward", (reward_information) => {
  // do changes here!
  console.log(reward_information);

  // TODO: show reward on seperate secured screen! And stop updates/refreshes of screen

})

// Bind Submit Event for Front Page Group Joiner / Group Starter / Group Resume
window.addEventListener("load",function(event) {
  $( "#start-group-form" ).on( "submit", function(e) {
      e.preventDefault();

      console.log(`Attempting to start a group!`)

      socket.emit('join-a-group');
  });

  $( "#join-group-form" ).on( "submit", function(e) {
      e.preventDefault();

      var group = $("#groupId").val();
      console.log(`Attempting to join ${ group }`)

      socket.emit('join-a-group', group);
  });
}, false);


server.js
//***SERVER.js***//
const express = require('express');
const app = require('express')();
const http = require("http").createServer(app);
const io = require('socket.io')(http);
const path = require('path')
const randomstring = require ('randomstring');
var room_size = 0; //current number of playing players

// Our Database
const boot_database = require('./app/database').db;

// configuration
const PORT = process.env.PORT || 5000;
const CONNECTION_STRING = process.env.DATABASE_URL;

// Startup Logic
app.use(express.static(path.join(__dirname, 'public')))
  .set('views', path.join(__dirname, 'views'))
  .set('view engine', 'ejs')
  .get('/', (request, response) => response.render('pages/index'))
  .get('/reward', (request, response)=> response.render('pages/reward'));

function getGameByCode(games_result, game_id) {
    for (var i = 0; i < games_result.length; i++) {
      if (games_result[i]["game_code"] == game_id) {
	  return games_result[i];
      };
    };
    console.log("Game ", game_id, "is invalid");
    return null;
}; // end getGameByCode

// SOCKET FUNCTIONAL CODE
async function configure_socketio(db_connection, games_result) {
   io.on("connection", (socket) => {
        //console.log(socket.id, "connected");
        socket.on('location-update', (...args) => {
           let location_update_query = `
              UPDATE player
              set location = ST_GeomFromText('POINT(${args[0]} ${args[1]})', 3857)
              WHERE id = '${socket.id}'
           `;
           db_connection.query(location_update_query).catch(err => console.log(err));
           //console.log(socket.id, "performed location update", args);
        }); // location-update

	socket.on('LOGTX', (clientlogdata) => {console.log(clientlogdata)}); //Receive console log data from client to drop on server console
		  //clientlogdata should always be in the format of `${socket.id} :-> log message`
		     
        socket.on('join-a-game', (chosen_game, callback) => {
            // Start Chosen Game ...
            var game_details = getGameByCode(games_result, chosen_game);
            if(game_details != null) {  // Confirm game exists, a non null valid = successful game choice
                var gamedesc = game_details["description"];
                console.log("Chosen Game is", chosen_game, ":", gamedesc); 
		socket.join("game-" + chosen_game);
                callback({ status: "Success", message: gamedesc }); //WHY DOES THIS FAIL AND CRASH THE DYNO? CAN WE CATCH IT?
            } else {
		var game_error = "Invalid Game Code: " + chosen_game + " Please try again!"    
                callback({ status: "Error", message: game_error });  // Invalid game, leave form on screen with error message below
            };
        }); // join-a-game
    }); // on connection

    // Respond to SOCKET ADAPTER EVENTS - a player or a room
    // Create - start a new game room
    io.of("/").adapter.on("create-room", (room) => {
       if (room.startsWith("game-")) {
          console.log("new game:", room.replace("game-", "")); // NEW GAME ROOM STARTED
        } else {
          console.log("new player:", room); // NEW PLAYER STARTED
        };
    }); // create-room

    // Join a game room - Inform users within game of changes (i.e. count of active players)
    io.of("/").adapter.on("join-room", (room, id) => {
      if(room !== id) {
        db_connection.query(`INSERT INTO player(id, room_id) VALUES('${id}', '${room.replace("game-", "")}')`).catch(err => console.log(err));
        room_size = io.sockets.adapter.rooms.get(room).size; // number of currently connected players to the game
        console.log(id, "joined", room, ",", room_size, "players online");
        var game_details = getGameByCode(games_result, room.replace("game-", ""));
	var gamedesc = game_details["description"];
	io.to(room).emit("room-update", room.replace("game-", ""), gamedesc, room_size);
        let game_code = room.replace("game-", "");
        io.to(id).emit("game-join");
      };
    }); // join-room

    // Leave a game room - Inform users within game of changes (i.e. count of active players)
    io.of("/").adapter.on("leave-room", (room, id) => {
      if(room !== id) {
        room_size = io.sockets.adapter.rooms.get(room).size; // number of currently connected players to the game
        console.log(id, "left", room.replace("game-", ""), room_size, "online");
	if (room_size==0) {console.log("All players have left game: ", room.replace("game-", ""))};
	io.to(room).emit("room-update", room.replace("game-", ""), 1, room_size);
        db_connection.query(`DELETE FROM player WHERE id = '${id}'`).catch(err => console.log(err));
      }
    }); // leave-room
}; // end of CONFIGURE_SOCKET

function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}; // end of DELAY

// UPDATE GAME: Primary game management coding
async function update_game(room, io, db_connection, games_result) {
    if(!io.sockets.adapter.rooms.has(room)) {
       return; // error check in case there is no current game (room) in motion
    };
    if(!room.includes("game-")) {
	     return; // error check in case a player ID gets confused in here
    };
	
   let game_code = room.replace("game-", "");
   //console.log("Playing game", game_code);
   let game_details = getGameByCode(games_result, game_code); // still needed here to populate the game_details variable again?

   let display_query = `
       SELECT pl.id, pl.room_id, pl.updated_at,
              wp.name, wp.radius, st_x(wp.location) as "x", st_y(wp.location) as "y", round(ST_DISTANCE(wp.location, pl.location) * 100000) as "distance"
       FROM player as pl, waypoint as wp
       WHERE wp.game_code = '${game_code}' AND pl.room_id = '${game_code}'
       ORDER BY pl.id
    `;
    let display_result = await db_connection.query(display_query);
    //console.table(display_result.rows);
    io.to(room).emit('display-update', display_result.rows);

    // How many waypoints are there in this game?
    let minimum_player_count = game_details["minimum_players"]

    // Determine whether they have "met the criteria" to succeed in the game!
    var within_radius = []; //lists all target waypoints that are occupied, uniquely
    var winning_player =[]; //lists all players occupying a target waypoint
    for (var i = 0; i < display_result.rows.length; i++) {
	var row = display_result.rows[i];
	var distance = row.distance;
	var radius = row.radius;
	if(distance != null && distance <= radius ) { // error check, should never be null at this point
		console.log(row.id, "is occupying waypoint", row.name);
		winning_player.push(row.id); //add a winning player if they are inside a waypoint 
		if ( !within_radius.includes(row.name) && distance <= radius ) { // only include a waypoint if not already occupied
			within_radius.push(row.name); 
		};
	};
    };
    console.log(game_code, ":", within_radius.length, "of", minimum_player_count, "waypoints occupied.", room_size, "Players online.");

    if (within_radius.length == minimum_player_count) { //If the number of occupied waypoints == the number required we have success
       	let reward = game_details["reward"];
	for (var p = 0; p < winning_player.length; p++) { //Emit the reward ONLY to an occupying player, not the whole room
	    console.log("REWARD goes to:", winning_player[p], reward); //this will display the actual reward data on to the server console log
	    io.to(winning_player[p]).emit('display-reward', reward);
	};
    }; // Send reward
}; // end of UPDATE_GAME

async function main() {
    // On SERVER Startup - Boot the database and delete all players left over from previous runs
    let connection = await boot_database(CONNECTION_STRING);
    let startingup = await connection.query("DELETE FROM player").catch(err => console.log(err));
    console.log("System startup");
    let games_result = await connection.query("SELECT game_code, description, reward, minimum_players FROM games").catch(err => console.log(err));
    await configure_socketio(connection, games_result.rows)
    // Commence listening for client connections
    http.listen(PORT, () => console.log(`listening on *:${ PORT }`));
	
    while ( 1 == 1 ) { // endless loop runs the games
      Array.from( io.sockets.adapter.rooms.keys() ).forEach(roomId => { // For each concurrently running game
         update_game(roomId, io, connection, games_result.rows);
      });
      await delay(10000); // wait 10 seconds between performing game updates
    }; // end of while loop
} // end of MAIN

main();  // RUN IT ALL!
